src/Controllers/DatabaseController.js

```diff
@@ -1390,6 +1410,42 @@ class DatabaseController {
     }
   }
 
+  addProtectedFields(
+    schema: SchemaController.SchemaController,
+    className: string,
+    query: any = {},
+    aclGroup: any[] = [],
+    auth: any = {}
+  ) {
+    const perms = schema.getClassLevelPermissions(className);
+    if (!perms) return null;
+
+    const protectedFields = perms.protectedFields;
+    if (!protectedFields) return null;
+
+    if (aclGroup.indexOf(query.objectId) > -1) return null;
+    if (
+      Object.keys(query).length === 0 &&
+      auth &&
+      auth.user &&
+      aclGroup.indexOf(auth.user.id) > -1
+    )
+      return null;
+
+    let protectedKeys;
+    [...(auth.userRoles || []), '*'].forEach(role => {
+      // If you are in multiple groups assign the role with the least protectedKeys.
```

Comment:
Good call on the warning. Length comparison seems dangerous. Idea: start with the base assumption that all protectedKeys are shielded, then iterate through each role, and each item in protectedKeys. For any fields that are in `protectedKeys`, but not `protectedFields[role]`, remove the key from the return value of `protectedKeys`

Reply:
Yeah - Much better idea. Thanks! Implemented in next commit. Hope your project works out.